MediaGalleryMetadata/Model/File/AddMetadata.php

```diff
@@ -7,96 +7,148 @@
 
 namespace Magento\MediaGalleryMetadata\Model\File;
 
-use Magento\Framework\Exception\FileSystemException;
 use Magento\Framework\Exception\LocalizedException;
-use Magento\MediaGalleryMetadataApi\Api\AddMetadataInterface;
 use Magento\MediaGalleryMetadataApi\Api\Data\MetadataInterface;
 use Magento\MediaGalleryMetadataApi\Model\FileInterface;
-use Magento\MediaGalleryMetadataApi\Model\FileReaderInterface;
-use Magento\MediaGalleryMetadataApi\Model\FileWriterInterface;
-use Magento\MediaGalleryMetadataApi\Model\MetadataWriterInterface;
+use Magento\Framework\Exception\ValidatorException;
+use Magento\MediaGalleryMetadataApi\Api\AddMetadataInterface;
+use Magento\MediaGalleryMetadataApi\Model\FileInterfaceFactory;
+use Magento\MediaGalleryMetadataApi\Model\ReadFileInterface;
+use Magento\MediaGalleryMetadataApi\Model\WriteFileInterface;
+use Magento\MediaGalleryMetadataApi\Model\WriteMetadataInterface;
 
 /**
- * Add metadata to the asset by path. Should be used as a virtual type with a file type specific configuration
+ * Write metadata to the asset file for all supportet types e.g IPTC, XMP ...
  */
 class AddMetadata implements AddMetadataInterface
 {
     /**
-     * @var MetadataWriterInterface[]
+     * @var array
      */
-    private $writers;
+    private $metadataWriters;
 
     /**
-     * @var FileReaderInterface
+     * @var FileInterfaceFactory
      */
-    private $fileReader;
+    private $fileFactory;
 
     /**
-     * @var FileWriterInterface
+     * @var array
      */
-    private $fileWriter;
+    private $fileReaders;
 
     /**
-     * @param FileReaderInterface $fileReader
-     * @param FileWriterInterface $fileWriter
-     * @param MetadataWriterInterface[] $writers
+     * @var array
      */
-    public function __construct(FileReaderInterface $fileReader, FileWriterInterface $fileWriter, array $writers)
-    {
-        $this->fileReader = $fileReader;
-        $this->fileWriter = $fileWriter;
-        $this->writers = $writers;
+    private $fileWriters;
+
+    /**
+     * @param FileInterfaceFactory $fileFactory
+     * @param array $metadataWriters
+     * @param array $fileReaders
+     * @param array $fileWriters
+     */
+    public function __construct(
+        FileInterfaceFactory $fileFactory,
```

Comment:
AddMetadata and ExtractMetdata have too much responsibility. I'd keep the current approach with virtual types